### ISSUE TITLE
no-unsafe-any should not complain when () => any is assigned to a Function

### ISSUE THREAD

**Issue by madelson** [palantir/tslint#4587]
### Bug Report

- __TSLint version__: 5.10
- __TypeScript version__: 3.0
- __Running TSLint via__:  Visual Studio

#### TypeScript code being linted

```ts
        const func = // strongly-typed function returning void passed in from elsewhere
	const x = () => func.apply(null, args);
	setTimeout(x, 100);
```

#### Actual behavior
Error: Unsafe use of expression of type 'any'

#### Expected behavior
No error. The fix as to as `as void` after the apply call. However, given that `setTimeout` takes `string | Function` I don't think I should have to specify a more specific return type to make tslint happy.

**Comment by JoshuaKGoldberg** [comment]
Hi @madelson! The `setTimeout` isn't what's triggering the rule; it's the `func.apply(null, args)`. Simplifying the code a bit:

```ts
const returnsVoid = (): void => { };
const returnsUnsafe = () => returnsVoid.apply(null);
                         // ~~~~~~~~~~~~~~~~~~~~~~~~
                         // ERROR: 2:17  no-unsafe-any  Unsafe use of expression of type 'any'.
```


If you enable [`--strictBindCallApply`](https://www.typescriptlang.org/docs/handbook/compiler-options.html), the function created by `.apply` will have the correct function signature, which should solve your issue, right?

**Comment by madelson** [comment]
@JoshuaKGoldberg I see what you're saying. I believe I still get the warning though with this code:

```
setTimeout(() => returnsVoid.apply(null), 100);
```

I'm not sure what the full intent of the unsafe-any rule is but neither of these feel right to me. In both cases we're not "using" the `any` value as a value of some type but are either continuing to declare it as `any` or coalescing it to an equivalent (e. g. `() => any` => `() => void` or `() => any` => `Function` both feel very "safe" to me). Thoughts?

**Comment by JoshuaKGoldberg** [comment]
Hmm, yes. You're correct that in _this instance_ the usage isn't truly unsafe. But, how could TSLint know the difference between this and other functions that do use the return type?

```ts
// ...what a terrible name!
const setTimeout = (callback: Function, timeout: number) => {
    window.setTimeout(
        () => {
            doSomethingVerySensitive(callback());
        },
        timeout);
}

const returnsVoid = (): void => {};

setTimeout(() => returnsVoid.apply(null), 100);
```

In that situation, there's really no clean way for the rule to know that the `Function`-typed function it's being passed to doesn't use its return value.

As for allowing `any` to be returned by a `() => any` into a location that receives `() => void`, that seems like it might be a reasonable option for the rule? The rule's goal is partially to discourage handling `any` types when more specific types will do (as a great counterpart to disabling your own `any` declarations). It'd be nice to get some more feedback in case there's more to it.

**Comment by madelson** [comment]
> But, how could TSLint know the difference between this and other functions that do use the return type?

I'm not sure what information TSLint is privy to from TSC, but my feeling is that this rule would be more useful if it were restricted to cases where `any` was being used in an unsafe way:

```
const naturalAny = returnsVoid.apply(null); // should be fine
takesAnAny(naturalAny); // should be fine
takesUnknown(naturalAny); // should be fine
const returnsAny = () => naturalAny; // should be fine
const returnsVoid: () => void = returnsAny; // should be fine
const returnsUnknown: () => unknown = returnsAny; // should be fine
const func: Function = returnsAny; // should be fine

takesANumber(naturalAny); // error! (cast to number)
naturalAny.foo(); // error! (cast to { foo(): void })
```

Curious whether you agree.

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
